libp2p/src/main/kotlin/io/libp2p/core/multiformats/Multiaddr.kt

```diff
@@ -61,7 +61,11 @@ data class Multiaddr(val components: List<MultiaddrComponent>) {
      * @throws IllegalArgumentException if existing component value doesn't match [value]
      */
     private fun withComponentImpl(protocol: Protocol, value: ByteArray?): Multiaddr {
-        val existingComponent = getFirstComponent(protocol)
+        val existingComponent = if (has(Protocol.P2PCIRCUIT)) {
+            split { it == Protocol.P2PCIRCUIT }.get(1).getFirstComponent(protocol)
```

Comment:
I think those existing `withComponent` methods hit their limitations with introducing `relay` protocol. I believe we would need to introduce a builder for `Multiaddr` which would be able to construct/modify components in a more explicit way.   
This could be addressed later in a separate PR